config/module.go

```diff
@@ -213,3 +214,59 @@ func (m Module) EvaluateExePath(packagesDir string) (string, error) {
 	}
 	return m.ExePath, nil
 }
+
+// EvaluateFirstRunPath returns absolute FirstRunPath from one of three sources (in order of precedence):
+// 1. if there is a meta.json in the exe dir, use that, except in local non-tarball case.
+// 2. if this is a local tarball and there's a meta.json next to the tarball, use that.
+// 3. otherwise use the exe path from config, or fail if this is a local tarball.
+// Note: the working directory must be the unpacked tarball directory or local exec directory.
+func (m Module) EvaluateFirstRunPath(packagesDir string) (string, error) {
+	firstRunDir, err := m.exeDir(packagesDir)
+	if err != nil {
+		return "", err
+	}
+	// note: we don't look at internal meta.json in local non-tarball case because user has explicitly requested a binary.
+	localNonTarball := m.Type == ModuleTypeLocal && !m.NeedsSyntheticPackage()
+	if !localNonTarball {
+		// this is case 1, meta.json in exe folder.
+		metaPath, err := utils.SafeJoinDir(firstRunDir, "meta.json")
+		if err != nil {
+			return "", err
+		}
+		_, err = os.Stat(metaPath)
+		if err == nil {
+			// this is case 1, meta.json in exe dir
+			meta, err := parseJSONFile[JSONManifest](metaPath)
+			if err != nil {
+				return "", err
+			}
+			firstRun, err := utils.SafeJoinDir(firstRunDir, meta.FirstRun)
+			if err != nil {
+				return "", err
+			}
+			// TODO: investigate
+			return filepath.Abs(firstRun)
+		}
+	}
+	if m.NeedsSyntheticPackage() {
+		// this is case 2, side-by-side
+		// TODO(RSDK-7848): remove this case once java sdk supports internal meta.json.
+		metaPath, err := utils.SafeJoinDir(filepath.Dir(m.ExePath), "meta.json")
+		if err != nil {
+			return "", err
+		}
+		meta, err := parseJSONFile[JSONManifest](metaPath)
+		if err != nil {
+			// note: this error deprecates the side-by-side case because the side-by-side case is deprecated.
+			return "", errors.Wrapf(err, "couldn't find meta.json inside tarball %s (or next to it)", m.ExePath)
+		}
+		firstRun, err := utils.SafeJoinDir(firstRunDir, meta.FirstRun)
+		if err != nil {
+			return "", err
+		}
+		// TODO: investigate
+		return filepath.Abs(firstRun)
+	}
```

Comment:
@abe-winter just to confirm my understanding of this code path - our code docs say the following about `NeedsSyntheticPackage`

> NeedsSyntheticPackage returns true if this is a local module pointing at a tarball

So basically this refers to module tarballs that are not coming from the registry - the user just has them available locally on the machine - is that correct?

Reply:
thanks! 

@cheukt to confirm, should this be a valid path to accessing a setup script? my thinking is yes but let me know if you disagree